File: server/features/mailing/mailing-core/src/main/java/de/bennyboer/kicherkrabbe/mailing/settings/MailgunSettings.java
package de.bennyboer.kicherkrabbe.mailing.settings;

import jakarta.annotation.Nullable;
import lombok.AllArgsConstructor;
import lombok.Value;
import lombok.With;

import java.util.Optional;

import static lombok.AccessLevel.PRIVATE;

@Value
@With(PRIVATE)
@AllArgsConstructor(access = PRIVATE)
public class MailgunSettings {

    @Nullable
    ApiToken apiToken;

    public static MailgunSettings of(@Nullable ApiToken apiToken) {
        return new MailgunSettings(apiToken);
    }

    public static MailgunSettings init() {
        return new MailgunSettings(null);
    }

    public Optional<ApiToken> getApiToken() {
        return Optional.ofNullable(apiToken);
    }

    public MailgunSettings updateApiToken(ApiToken apiToken) {
        return withApiToken(apiToken);
    }

    public MailgunSettings clearApiToken() {
        if (getApiToken().isEmpty()) {
            throw new MailgunApiTokenAlreadyClearedException();
        }

        return withApiToken(null);
    }

}

File: frontend/management/src/app/modules/admin/modules/mailing/model/mailgun-settings.ts
import { none, Option, someOrNone, validateProps } from '@kicherkrabbe/shared';

export class MailgunSettings {
  readonly apiToken: Option<string>;

  private constructor(props: { apiToken: Option<string> }) {
    validateProps(props);

    this.apiToken = props.apiToken;
  }

  static of(props: { apiToken?: string }): MailgunSettings {
    return new MailgunSettings({
      apiToken: someOrNone(props.apiToken),
    });
  }

  updateApiToken(apiToken: string): MailgunSettings {
    return new MailgunSettings({
      apiToken: someOrNone(apiToken),
    });
  }

  clearApiToken(): MailgunSettings {
    return new MailgunSettings({
      apiToken: none(),
    });
  }
}
